yellow-paper/docs/public-vm/execution.md

```diff
@@ -0,0 +1,190 @@
+
+## Halting
+
+A context's execution can end with a **normal halt** or **exceptional halt**. A halt ends execution within the current context and returns control flow to the calling context.
+
+### Normal halting
+
+A normal halt occurs when the VM encounters an explicit halting instruction ([`RETURN`](./instruction-set#isa-section-return) or [`REVERT`](./instruction-set#isa-section-revert)). Such instructions consume gas normally and optionally initialize some output data before finally halting the current context's execution.
+
+```
+machineState.l1GasLeft -= instr.l1GasCost
+machineState.l2GasLeft -= instr.l2GasCost
+machineState.daGasLeft -= instr.daGasCost
+results.reverted = instr.opcode == REVERT
+results.output = machineState.memory[instr.args.retOffset:instr.args.retOffset+instr.args.retSize]
+```
+
+> Definitions: `retOffset` and `retSize` here are arguments to the [`RETURN`](./instruction-set/#isa-section-return) and [`REVERT`](./instruction-set#isa-section-revert) instructions. If `retSize` is 0, the context will have no output. Otherwise, these arguments point to a region of memory to output.
+
+> `results.output` is only relevant when the caller is a contract call itself. In other words, it is only relevant for [nested contract calls](./nested-calls). When an [initial contract call](./context#initial-contract-calls) (initiated by a public execution request) halts normally, its `results.output` is ignored.
+
+### Exceptional halting
+
+An exceptional halt is not explicitly triggered by an instruction but instead occurs when an exceptional condition is met.
+
+When an exceptional halt occurs, the context is flagged as consuming all of its allocated gas and is marked as `reverted` with no output data, and then execution within the current context ends.
+
+```
+machineState.l1GasLeft = 0
+machineState.l2GasLeft = 0
+machineState.daGasLeft = 0
+results.reverted = true
+// results.output remains empty
+```
+
+The AVM's exceptional halting conditions area listed below:
+
+1. **Insufficient gas**
+    ```
+    assert machineState.l1GasLeft - instr.l1GasCost > 0
+    assert machineState.l2GasLeft - instr.l2GasCost > 0
+    assert machineState.daGasLeft - instr.l2GasCost > 0
+    ```
+1. **Invalid instruction encountered**
+    ```
+    assert environment.bytecode[machineState.pc].opcode <= MAX_AVM_OPCODE
+    ```
+1. **Jump destination past end of bytecode**
+    ```
+    assert environment.bytecode[machineState.pc].opcode not in {JUMP, JUMPI, INTERNALCALL}
+        OR instr.args.loc < environment.bytecode.length
+    ```
+1. **Failed memory tag check**
+    - Defined per-instruction in the [Instruction Set](./instruction-set)
+1. **Maximum memory index ($2^{32}$) exceeded**
```

Comment:
Maybe it's clearer to call it "Out of bounds memory access". This not only includes wrong offsets, but also slices (offset+size).